### DISCUSSION TITLE
`mcmap` browser support for large maps.

### DISCUSSION THREAD

**Discussion by spoutn1k** [spoutn1k/mcmap#91] [Ideas]
This thread is made to discuss the implementation of a new browser-based solution to view large `mcmap` maps.

A legacy [OpenLayers](https://openlayers.org/) solution existed in the `contrib` folder - and I added a new, [Leaflet](https://leafletjs.com/) based one. A project called mcmapDZ used to work with `hdmake`, but both projects seem lost to time.

Let's discuss the available implemetations and what is expected from a valid solution, along with what a workflow would look like.

List of target features:

- [x] Smooth movement
- [x] Zoom
- [x] Markers
- [x] Marker toggle in window

**Comment by DrakorthaP** [comment]
Smooth movement and zoom.

Markers switch on/off within the browser window.

**Comment by spoutn1k** [comment]
Had to find some time, but the zoom and smooth scrolling are done. Can you build locally ? I pushed the changes to the `web-map` branch. Now we need to figure out the markers and a reliable hosting solution.

**Reply by DrakorthaP** [reply]
Thanks for the update! What did you mean by build locally?

**Reply by spoutn1k** [reply]
Like, how are you using `mcmap` ? Can you build the project on your machine ? To see the changes, you will need to clone the repo and build a specific branch.

**Comment by spoutn1k** [comment]
How do you store your markers ? Do you usually pass them on the command line/UI or do you have a file with all of them defined ?

I am thinking of implementing a JSON file input with a list of locations in minecraft coordinates, that will get output in the map metadata with their coordinates updated for the render. Something like:

```json
{
  "worldX": 1789,
  "worldY": -892,
  "color": "#whatever",
  "text": "This marks my secret entrance",
  "icon": "file.png"
}
```

Leaflet allows the use of custom icons, are you interested in that ?

**Reply by DrakorthaP** [reply]
I can't say I understand the technicals of what's happening under the hood. A lot of what you are saying goes straight over my head.

I've always used mcmapdz's UI to create renders of my Minecraft world. This UI had labels and everything built into the UI for easy use.

**Comment by spoutn1k** [comment]
In a programming tour de force, I got the markers to work in less than 2 hours !

I am not a UI designer. For now, marker support in `mcmap` will pass through JSON files. You will have to create a file with the following format:

```json
[{
    "x": 1247,
    "y": 72,
    "z": -622,
    "title": "This is my base."
}, {
    "x": 0,
    "y": 100,
    "z": 0,
    "title": "The ender portal !"
}]
```

The `x,y,z` fields are the coordinates of the marker you would like to place. The `title` field will appear on the map when clicked.

Once done, they will appear on the map !

![Screenshot from 2023-01-29 22-38-21](https://user-images.githubusercontent.com/22240065/215405771-82ac9ed2-2bad-4bb0-bba3-a73b9417d9de.png)
![Screenshot from 2023-01-29 22-38-04](https://user-images.githubusercontent.com/22240065/215405776-b68f322f-29e6-4540-b9eb-7e7c1b7d8d4c.png)

You can also disable all markers with a menu:
![Screenshot from 2023-01-29 22-47-16](https://user-images.githubusercontent.com/22240065/215406962-2f584794-1526-4d5f-9572-923adf8a491a.png)
![Screenshot from 2023-01-29 22-46-36](https://user-images.githubusercontent.com/22240065/215406834-8460eb68-5800-4be2-ab14-f86fb347a3a2.png)

**Reply by DrakorthaP** [reply]
This is brilliant! I will follow your progress :)

**Comment by dferreras439** [comment]
Hello! I am also interested in web browser usage, and am very impressed with the marker system! Thank you!

As for increased leaflet performance for large maps, could I suggest adding jpeg output to mcmap? I would like to use these compressed jpegs as LODs and have leaflet.js render them for when you're zoomed out -- [example](https://github.com/Esri/esri-leaflet/issues/237#issuecomment-41826334), thus increasing performance for large maps.: 
```
L.TileLayer("xxxxxx", {
  minZoom: 11,
  maxZoom: 21
}).addTo(map);
```

I suspect the map that @DrakorthaP posted as an example, [link](https://qkzskrbbwbkyw3lacbvz9w.on.drv.tw/www.taw.map/), uses some sort of jpeg LOD too.

**Reply by spoutn1k** [reply]
I am not familiar with LODs, could you elaborate ?

I made `mcmap` spit out downscaled versions of the map for other zoom levels, so large maps are rendered much more efficiently from that, but I am open to implement new ways of boosting it - when I'll have time.

**Reply by dferreras439** [reply]
I didn't know mcmap already produced downscaled versions! I had not looked into the commit history. I'll build it later today to try it out :) 

By LODs I meant just that, downscaled 'level-of-detail' versions - arguably I used the term incorrectly as the 'LOD' refers to 3d objects rendered from afar using low quality versions. But the concept is the same just in 2d, pardon the confusion. 

In my comment I meant using a harsher, lossier compression algorithm such as jpeg for the downscaled tiles as opposed to the traditionally lossless png, for really really big maps.

**Reply by spoutn1k** [reply]
Oh it is fresh as of yesterday. Thanks for clarifying - as `mcmap` was able to output downscaled tiles that leaflet understood, I'm going to put jpegs on the backburner. Please test and let me know ! `mcmap-gui` should be broken with those new changes. Use the `-marker` option to pass a marker file.

**Reply by dferreras439** [reply]
Built it on debian 11, works nicely! Didn't try it on a big map, just in case, but will do at some point. Cheers!
